src/components/ComposerQuill/composer.js

```diff
@@ -100,70 +101,84 @@ class Composer extends React.Component {
-    // we can ignore p tags because if there are no other element tags, we can just display the original text instead
-    const shortened = parsed.replace(/<\/?p>/g, '');
-
-    // checks if we have any other html tags. this would indicate there were markdowns in the original text
-    // this should not match <br /> tags
-    const hasMarkdown = /<.+?>.+<\/.+?>/.test(shortened);
-
-    const object = {displayName: original.trim()};
-
-    // if there are no markdowns, then we only need to send the displayName for the activity object
-    // if there are markdowns, then strip the markdowns from the text for displayName
-    // and send content with the html tags
-    if (hasMarkdown) {
-      // removes all html tags
-      const stripped = shortened.replace(/<.+?>/g, '');
-
-      object.displayName = stripped;
-      object.content = parsed;
-    }
+    const {onError, send} = this.props;
+
+    try {
+      // get the text from the composer as-is and a sanitized version
+      // check the sanitized version if there are markdown or mentions in it
+      // if there are, then we will parse and use the sanitized version
+      // otherwise we will send the original without the content property
+      const {original, sanitized} = getQuillText(this.quill);
+
+      // converts text from markdown to html
+      // element tags will have new lines after them which we don't want so we remove them here too
+      // new lines in the text will be represented with a br tag so no need to worry about them
+      const parsed = md.render(sanitized).replace(/\n/g, '');
+
+      // after parsing, text will have the p tags around it, remove them so we can check if there are other html tags present
+      // we can ignore p tags because if there are no other element tags, we can just display the original text instead
+      const shortened = parsed.replace(/<\/?p>/g, '');
+
+      // checks if we have any other html tags. this would indicate there were markdowns in the original text
+      // this should not match <br /> tags
+      const hasMarkdown = /<.+?>.+<\/.+?>/.test(shortened);
+
+      const object = {displayName: original.trim()};
+
+      // if there are no markdowns, then we only need to send the displayName for the activity object
+      // if there are markdowns, then strip the markdowns from the text for displayName
+      // and send content with the html tags
+      if (hasMarkdown) {
+        // removes all html tags
+        const stripped = shortened.replace(/<.+?>/g, '');
+
+        object.displayName = stripped;
+        object.content = parsed;
+      }
 
-    send(object);
-    // clear the composer and reset the draft
-    this.quill.setText('');
-    this.saveToDraft();
+      send(object);
+      // clear the composer and reset the draft
+      this.quill.setText('');
+      this.saveToDraft();
+    } catch (e) {
```

Comment:
do we need to wrap these around try/catch? I thought the componentDidCatch would catch these. If we are catching the error here, I don't think componentDidCatch will catch the error. What will the message composer do in these cases if the errors are caught?

Reply:
componentDidCatch doesn't catch errors from this functions. Maybe it's cause i'm passing these functions to quill to call instead? 

The function will fail and stop, but the composer doesn't crash. For example, this just wouldn't send the message. Then the catch would tell the client who would log it and send a metric that this function failed.

Reply:
will the user understand what happened though? for example, they entered something incorrect, this crashed and caught the error. the message will still be in the message composer? but will the user know what to do or why s/he wasn't able to send it and what can s/he do next?

Reply:
i think if you let the message composer error boundary catch it, you can get the same information on where the crash happened right? do we need it to be more granular?

Reply:
Unless the user looks in the console to see the error report, then no they won't really know what happened. For example in this case it would fail to send the message and instead enter a new line. It's because this error out and just stopped this function instead of crashing the whole composer. 

If we don't add this, the message composer from our client won't catch it. That boundary only catches full-on crashes of the composer, whereas this was just an error that caused the function to fail. If we didn't catch it then it would only display the error message to the console. This at least sends a metric to us so we know that something happened.

Reply:
oo intersting. i thought if there was an error somewhere in the component, it would bubble up to the componentDidCatch (if it wasn't wrapped in a try/catch). if that's not the case, yeah i agree it's better to send a metric in that case then

Reply:
I did too! But the only way i've been able to get it to full-on crash was by causing an error in constructor, one of the life cycle methods, or render. Though I did seem to do one through a utils file that it caught, but when I tried to do it again later it didn't and went through the error instead..

That said after talking with Chris we decided to just wrap a few functions for as a template. It'll tell you which function from the ComposerQuill file it came from, but it's still minified and i haven't been able to get it to show the name of the function if it came from like the utils file or something.